src/charm.py

```diff
@@ -103,7 +117,119 @@ def _on_stop(self, _: ops.StopEvent) -> None:
     def _on_upgrade_charm(self, _: ops.UpgradeCharmEvent) -> None:
         """Handle upgrade-charm."""
         self.unit.status = ops.MaintenanceStatus("Upgrading dependencies")
-        self.bind.prepare(self.unit.name)
+        self.bind.setup(self.unit.name)
+
+    def _on_leader_elected(self, _: ops.LeaderElectedEvent) -> None:
+        """Handle leader-elected event."""
+        # We check that we are still the leader when starting to process this event
+        if self.unit.is_leader():
+            self._become_active()
+
+    def _on_peer_relation_departed(self, _: ops.RelationDepartedEvent) -> None:
+        """Handle the peer relation departed event."""
+        # We check that we are still the leader when starting to process this event
+        if self.unit.is_leader():
+            self._become_active()
+
+    def _become_active(self) -> bool:
+        """Set the current unit as the active unit of the charm.
+
+        Returns:
+            True if the charm is effectively the new active unit.
+        """
+        active_unit_ip = self._active_unit_ip()
+
+        assert self.peer_relation is not None  # nosec
```

Comment:
If there is no peer, should the charm go to `WaitingStatus`? I'm not sure if the ErrorState is intentional, if so, it might be worth logging what is expected to help debug :)

Reply:
This is handled line 92 in the same file. It was a BlockedStatus but I changed it to WaitingStatus: you're right that it is more appropriate.